lib/galaxy/webapps/galaxy/services/datasets.py

```diff
@@ -493,6 +529,36 @@ def converted(
         hda = self.hda_manager.get_accessible(decoded_id, trans.user)
         return self.hda_serializer.serialize_converted_datasets(hda, "converted")
 
+    def delete_batch(
+        self,
+        trans: ProvidesHistoryContext,
+        payload: DeleteDatasetBatchPayload,
+    ) -> DeleteDatasetBatchResult:
+        """
+        Deletes or purges a batch of datasets.
+        Warning: only the ownership of the dataset and upload state for HDAs is checked, no other checks or restrictions are made.
+        """
+        success_count = 0
+        errors = []
+        for dataset in payload.datasets:
+            try:
+                decoded_dataset_id = self.decode_id(dataset.id)
+                manager = self.dataset_manager_by_type[dataset.src]
+                dataset_instance = manager.get_owned(decoded_dataset_id, trans.user)
+                if dataset.src == DatasetSourceType.hda:
+                    self.hda_manager.error_if_uploading(dataset_instance)
+                if payload.purge:
+                    manager.purge(dataset_instance, flush=False)
```

Comment:
The purge especially could take a while. I think this should ideally be an asynchronous celery endpoint.